Keep pg_rewind log if incremental gprecoverseg fails.

paul-guo- opened the pull request:
Previously if incremental gprecoverseg fails we have to rerun with -v
to get the pg_rewind log for debugging. This is not convenient and also
sometimes the log output is not complete in the gprecoverseg log somehow.
Let's keep the pg_rewind log if incremental gprecoverseg fails.

asimrp reviewed (approved):
Thank you!

asimrp commented:
Exit status of rewind command could be checked and the command output be preserved in a separate file.  This appears more complex that your trick here, because multiple pg_rewind commands need to be run within a single gprecoverseg invocation.

paul-guo- commented:
> Exit status of rewind command could be checked and the command output be preserved in a separate file. This appears more complex that your trick here, because multiple pg_rewind commands need to be run within a single gprecoverseg invocation.

Hi Asim, the `&& touch` code implies exit status checking so we only keep failed pg_rewind log.

ashwinstar commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -433,7 +433,7 @@ def __init__(self, name, target_host, target_datadir,
         # file exists in target data directory because the target instance can
         # be started up normally as a mirror for WAL replication catch up.
-        rewind_cmd = '[ -f %s/standby.signal ] || PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (target_datadir, source_server, target_datadir)
+        rewind_cmd = 'rm -f /tmp/gp.recovery.ok.dbid%s; if [ ! -f %s/standby.signal ]; then PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (target_dbid, target_datadir, source_server, target_datadir)
```
where did "/tmp" come from? I feel super uncomfortable we reading or writing anything to "/tmp" directory for tool to operate. We should be leveraging data directory or log directory for any such purpose but nothing other than that.

paul-guo- replied:
Yes later yesterday I also felt hardcoding /tmp is not a good idea so using `gplog.get_logger_dir()` (typically /home/gpadmin/gpAdminLogs) now.

ashwinstar commented:
How would user become aware of the failure reason, will the same be logged in gprecoverseg log file?

dkrieger-pvtl reviewed (commented):
It took me a little while to see the mechanism of this PR: that if `GP_LOG_DIR/recovery.ok.dbidXX` exists, then `pg_rewind` worked and hence we can delete the progress file and this marker file.  Otherwise, we need to keep this marker file.  Since it took me a little bit to figure out, I think a comment is in order in the code to explain this semantic.

Re-runs of pg_rewind are fine as you delete this marker file on each run. 

Note that the progressFile logs are in fact on the segment hosts, not the master host, which is not ideal.  But aggregating them on the master isn't part of this PR.

dkrieger-pvtl commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -433,7 +433,7 @@ def __init__(self, name, target_host, target_datadir,
         # file exists in target data directory because the target instance can
         # be started up normally as a mirror for WAL replication catch up.
-        rewind_cmd = '[ -f %s/standby.signal ] || PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (target_datadir, source_server, target_datadir)
+        rewind_cmd = 'rm -f %s/recovery.ok.dbid%s; if [ ! -f %s/standby.signal ]; then PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (gplog.get_logger_dir(), target_dbid, target_datadir, source_server, target_datadir)
```
perhaps we want to use the same format as the progress file to make things clearer:

```
            self.progressFileKeepMarkerFile = '%s/pg_rewind.%s.dbid%s.out.keep' % (gplog.get_logger_dir(),
                                                                timeStamp,
                                                                targetSegment.getSegmentDbId())
```

paul-guo- replied:
Fixed. Seems better to use progressFile.success as the marker file.

paul-guo- commented:
> It took me a little while to see the mechanism of this PR: that if `GP_LOG_DIR/recovery.ok.dbidXX` exists, then `pg_rewind` worked and hence we can delete the progress file and this marker file. Otherwise, we need to keep this marker file. Since it took me a little bit to figure out, I think a comment is in order in the code to explain this semantic.

Added the below comment in the code.
```
+        # touch a marker file to indicate pg_rewind success. If pg_rewind fails,
+        # we keep the progress file (pg_rewind log) for debugging, else we
+        # delete log file.
+        rewind_cmd = rewind_cmd + ' && touch %s.success; fi' % (pipes.quote(progress_file))
```

dkrieger-pvtl commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -513,7 +513,7 @@ def __getProgressAndRemoveCmds(self, progressFile, targetSegmentDbId, targetHost
 
         removeCmd = base.Command("remove file",
-                                 "rm -f %s" % pipes.quote(progressFile),
+                                 "[ ! -f %s.success ] || rm -f %s.success %s" % (pipes.quote(progressFile), pipes.quote(progressFile), pipes.quote(progressFile)),
```
nit: this is probably cleaner:

```python3
"[ ! -f {0}.success ] || rm -f {0}.success {0}".format(pipes.quote(progressFile))
```

dkrieger-pvtl commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -443,6 +443,11 @@ def __init__(self, name, target_host, target_datadir,
+
         self.cmdStr = rewind_cmd
 
         Command.__init__(self, name, self.cmdStr, ctxt, target_host)
```
We have "progressFile.success" three times used and it makes it a bit harder to see...perhaps do this in this file:

```python3
    @staticmethod
    def getSuccessFilename(progress_file):
        return progress_file + '.success'
```

and then the accesses look like:

```python3
s = 'use this file %s' % SegmentRewind.getSuccessFilename(progress_file)
```

Then you can add a block comment on that static method explaining how it is used.

paul-guo- replied:
Note `__getProgressAndRemoveCmds` is used by pg_basebackup also. I thought about adding a new function to get the marker file but that way we need to populate that for full recovery (pg_basebackup) also. pg_basebackup log seems to be kept after full recovery fails already so adding the marker file for full recovery would confuse people (pg_basebackup log keeping is not done via the marker file).

dkrieger-pvtl replied:
ugh...you are right.  I forgot about that dependency.

ashwinstar replied:
Curious to know why logic (log file retention and display and such) differs between full `pg_basebackup` and incremental `pg_rewind` recovery from this perspective?

dkrieger-pvtl replied:
@paul-guo-   Are you sure the `pg_basebackup` files are kept?  https://github.com/greenplum-db/gpdb/blob/master/gpMgmt/bin/gppylib/operations/buildMirrorSegments.py#L666 suggests we remove the progress files for basebackup in the same way we do for rewind.

Perhaps we want to implement this fix for both?

ashwinstar replied:
I would really like behavior be same for FULL and INCREMENTAL (we need to make it easy for customers/support/us)

ashwinstar replied:
This comment was marked as resolved without answering the asked questions. Please I would really like to know the answer to asked question. Is the behavior same for FULL and INCREMENTAL? What are the differences? Why such a change is not required for FULL?

dkrieger-pvtl replied:
> This comment was marked as resolved without answering the asked questions. Please I would really like to know the answer to asked question. Is the behavior same for FULL and INCREMENTAL? What are the differences? Why such a change is not required for FULL?

@ashwinstar  I am pretty sure the behavior is different; from code inspection FULL recovery always removes the log file on the segments, even upon failure.

paul-guo- replied:
Oh, I guess the discussion was marked as resolved before the new comments so I missed the new ones? Anyway I double-confirmed via testing a demo cluster that pg_basebackup log is kept is pg_basebackup fails (non-zero return value) else the log is removed. The related code implementation is different.

paul-guo- replied:
Ignore my words please. The test was based on the new code.

dkrieger-pvtl reviewed (commented):
I think the change is good, but since gprecoverseg contains some very long functions and deep call stacks, I'd like us to make sure code we add is especially clean.  My suggestions might not be the cleanest, but are initial suggestions.

shivzone commented:
> How would user become aware of the failure reason, will the same be logged in gprecoverseg log file?

At the very least we should point to the pg_rewind log files to the end users along with the failed segments. @dkrieger-pvtl is looking into the this as part of improvements to gprecoverseg

ashwinstar commented:
> > How would user become aware of the failure reason, will the same be logged in gprecoverseg log file?
> 
> At the very least we should point to the pg_rewind log files to the end users along with the failed segments. @dkrieger-pvtl is looking into the this as part of improvements to gprecoverseg

- Are these files present on coordinator or individual segments?
- I think we need to ease out the experience for detecting the failure of gprecoverseg
- Even to support the product, mostly will need access to these files along with gprecoverseg log file, so whatever way we can make it simpler/easier/intuitive would be good

dkrieger-pvtl commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -433,7 +433,7 @@ def __init__(self, name, target_host, target_datadir,
         # file exists in target data directory because the target instance can
         # be started up normally as a mirror for WAL replication catch up.
-        rewind_cmd = '[ -f %s/standby.signal ] || PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (target_datadir, source_server, target_datadir)
+        rewind_cmd = 'rm -f %s.success; if [ ! -f %s/standby.signal ]; then PGOPTIONS="-c gp_role=utility" $GPHOME/bin/pg_rewind --write-recovery-conf --slot="internal_wal_replication_slot" --source-server="%s" --target-pgdata=%s' % (pipes.quote(progress_file), target_datadir, source_server, target_datadir)
```
Note the progressFile has a timestamp in it, so`rm -f progressFile.success` will likely not actually have any effect.  The progressFile looks like `pg_rewind.20210518_120012.dbid1.out`, and the one for this current run won't match that from the previous run.

This is probably ok, as keeping the logs is likely a good idea.

paul-guo- replied:
Good catch. We may just skip the rm operation here since in most cases the file has been removed by gprecovrseg. I added here just because I worry about an existing file affects the following logic (i.e. pg_rewind fails but the pg_rewind log is still removed due to this). Now that we know the success file name varies in each run so there's no need to worry about that.

dkrieger-pvtl commented:
> * Are these files present on coordinator or individual segments?

The progressFiles, currently only containing the result of `pg_rewind --progress`, are on the segment host where the incremental recovery is run.

shivzone commented:
> > > How would user become aware of the failure reason, will the same be logged in gprecoverseg log file?
> > 
> > 
> > At the very least we should point to the pg_rewind log files to the end users along with the failed segments. @dkrieger-pvtl is looking into the this as part of improvements to gprecoverseg
> 
> * Are these files present on coordinator or individual segments?
> * I think we need to ease out the experience for detecting the failure of gprecoverseg
> * Even to support the product, mostly will need access to these files along with gprecoverseg log file, so whatever way we can make it simpler/easier/intuitive would be good

Currently these relevant log files are on individual segments and not aggregated on the coordinator.

paul-guo- commented:
@dkrieger-pvtl I've addressed all your comments and modified the code accordingly. Do you and other guys think the PR is ready to go (other issues discussed in the thread could be in other PRs)?

ashwinstar reviewed (approved):
I would let @dkrieger-pvtl provide the final approval but I am removing my changes requested flag and approving the same.

This PR can go in though I would really like follow-up to address the usability aspect of the same. As for support and customers, locating the file on individual segment gets little tedious for utilities (where known convention has been on coordinator the information is available). Plus, customers might not know where to look for failure reason. So, we definitely need to ease that out. I think that's something CM team can carry forward @Nirali13 from here.

dkrieger-pvtl reviewed (approved):
LGTM...this is exactly the desired usecase.  Upon failure, the log/log.success file will be in the gplog dir for analysis on the segment hosts.

The implementation of this is heavily constrained by our current progress reporting style, which looks at the tail of a file on the segments to report the progress on the coordinator.  That's a much heavier lift to change but it might be time to revisit that in a follow up.

paul-guo- commented:
CI fails since it introduces regression in pg_basebackup related tests. I reverted the code and will recreate the new PR later.

paul-guo- commented:
I created a new PR https://github.com/greenplum-db/gpdb/pull/12032. The new PR seems to be a better fix and should have no regression for pg_basebackup.
